hphp/runtime/vm/jit/vasm-util.cpp

```diff
@@ -29,7 +29,12 @@ bool is_nop(const copy& i) { return i.s == i.d; }
 bool is_nop(const copy2& i) { return i.s0 == i.d0 && i.s1 == i.d1; }
 
 // movb r,r is a nop, however movl is not since it zeros upper bits.
-bool is_nop(const movb& i) { return i.s == i.d; }
+bool is_nop(const movb& i) {
+  // On PPC64 movb r,r cannot be considered a trivial nop since it zeros
```

Comment:
@racardoso—The specification for _every_ vasm instruction that has a Vreg8 src is that the upper bits of that src are considered killed when we hit the instruction (before it's executed).  For a Vreg8 dst, the upper bits are not well-defined once the instruction finished executed.

So, `movb{r, r}` takes `r := [56 bits of anything][8 defined bits]` and redefines it as `r := [56 bits of anything][the same 8 defined bits]`.  This is clearly a no-op.  It doesn't matter how you implement `movb{}` under the hood; by its vasm specification, a `movb{r, r}` is a no-op.

And you understood me correctly.  More generally, you can never use a Vreg defined with one width (e.g., Vreg8) in an instruction that expects a Vreg of a different width (e.g., Vreg64).  So, the result of any `b` instruction can only be used in instructions which take a Vreg8 (or an unconstrained Vreg).  If you want to use it for, e.g., an `addq` later or something, you need to do an intermediate `movzbq{}`.  Vreg8 and Vreg64 are not compatible types—much like how you can't use a `struct foo {};` as a `struct bar{};`, you cannot define a Vreg8 and then use it as a Vreg64.